### PR TITLE
GSLUX-537: Display internal metadata

### PR DESCRIPTION
**Pull request by tkohr**
### JIRA issue

https://jira.camptocamp.com/browse/GSLUX-537

### Description

PR displays layer metadata, when clicking on the info button within the

- catalog
- layer manager
- remote layers

The goal of this PR is to handle internal layers. A WMS helper has already been added and will be finished along with WMTS handling in a separate PR.

To-dos for this PR remain:

- [x] UI refinement
- [x] error handling
- [x] accessibility
- [x] add more tests
- [x] potentially refactoring composables

To-dos for separate PRs:

- finish querying metadata from external services (WMS, WMTS)
- allow URL configuration (using config provided by https://map.geoportail.lu/dynamic.json?interface=main&path=%2F)
- implement a generic modal to display information

### CODE REVIEW

File: src/components/catalog/catalog-tree.vue [hunk 1]
```diff
@@ -43,13 +47,41 @@ function toggleParent(node: LayerTreeNodeModel) {
 function toggleLayer(node: LayerTreeNodeModel) {
   layers.toggleLayer(parseInt(node.id, 10), !node.checked)
 }
+
+const { i18next } = useTranslation()
```
Comment (fgravin) [inline src/components/catalog/catalog-tree.vue:51]: Typically, all this section is a composable.
Shared in the tree and the layer manager

File: src/components/layer-manager/layer-manager.vue [hunk 2]
```diff
@@ -51,12 +56,39 @@ function toggleInfoLayer() {
...
+  metadata.value = undefined
+}
+
 function toggleEditionLayer() {
   console.info('clickEdit to implement')
 }
 </script>
 
 <template>
+  <layer-metadata
```
Comment (fgravin) [inline src/components/layer-manager/layer-manager.vue:87]: sounds weird that this component exists twice in the application

File: src/components/layer-manager/layer-manager.vue [hunk 3]
```diff
@@ -51,12 +56,39 @@ function toggleInfoLayer() {
   console.info('clickInfo to implement')
 }
 
+const { i18next } = useTranslation()
```
Comment (fgravin) [inline src/components/layer-manager/layer-manager.vue:59]: Don't really see why is this logic is here

File: src/components/remote-layers/remote-layers.vue [hunk 4]
```diff
@@ -120,9 +123,36 @@ function toggleLayer(node: LayerTreeNodeModel) {
     }
   }
 }
+
+const { i18next } = useTranslation()
```
Comment (fgravin) [inline src/components/remote-layers/remote-layers.vue:127]: 3rd times the same code

File: src/composables/common/formatting.composable.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+export default function useFormatting() {
```
Comment (fgravin) [inline src/composables/common/formatting.composable.ts:1]: not a composable at all

File: src/composables/layer-metadata/layer-metadata.composable.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  getResponsibleParty,
+  isoLang2To3,
+  stringToHtml,
+} from './layer-metadata.utils'
+import useWmsHelper from '../common/wms-helper.composable'
+import useThemes from '../themes/themes.composable'
+
+export default function useLayerMetadata() {
+  // TODO: get urls from a config
+  const geonetworkBaseUrl = 'https://geocatalogue.geoportail.lu/geonetwork/srv'
```
Comment (fgravin) [inline src/composables/layer-metadata/layer-metadata.composable.ts:16]: we should try to gather all the settings things to be sure not to forget anyone with hard coded values

File: src/composables/layer-metadata/layer-metadata.composable.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    let localMetadata
+    let metadata
+    if (layer) {
+      localMetadata = layer.metadata
+
+      const currentLanguage = i18next.language
+
+      metadata =
+        layer.metadata &&
+        layer.metadata.metadata_id &&
```
Comment (fgravin) [inline src/composables/layer-metadata/layer-metadata.composable.ts:39]:
```suggestion
        layer.metadata?.metadata_id &&
```

File: src/composables/layer-metadata/layer-metadata.composable.ts [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+      const legendHtml =
+        legendName &&
+        (await getLegendHtml(
+          legendBaseUrl,
+          legendName,
+          layerId,
+          currentLanguage
+        ))
+      metadata = legendHtml
```
Comment (fgravin) [inline src/composables/layer-metadata/layer-metadata.composable.ts:62]: factorize this a bit better

File: src/composables/layer-metadata/layer-metadata.composable.ts [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+import {
+  getMetadataLinks,
+  getResponsibleParty,
+  isoLang2To3,
+  stringToHtml,
+} from './layer-metadata.utils'
+import useWmsHelper from '../common/wms-helper.composable'
+import useThemes from '../themes/themes.composable'
+
+export default function useLayerMetadata() {
```
Comment (fgravin) [inline src/composables/layer-metadata/layer-metadata.composable.ts:14]: should be tested a little bit, many object manipulation in there.

**Review by fgravin — CHANGES_REQUESTED** [review]
Thanks @tkohr 
I think the logic on metadata fetching and panel hydration is correct.

Some thinking about the architecture of the metadata system though.
Why not just one `layer-metadata` component which is binded to a state which owns the `layerId` of the metadata to display ?
At least, the code should not be repeated 3 times and should be a composable in itself. But I don't even see the point to have the component present 3 times so it could be managed a bit simpler IMO.

File: src/composables/layer-metadata/layer-metadata.composable.ts (hunk 6 above)
Comment (tkohr) [reply src/composables/layer-metadata/layer-metadata.composable.ts]: yes, I've created a ticket for that

File: src/composables/layer-metadata/layer-metadata.composable.ts (hunk 8 above)
Comment (tkohr) [reply src/composables/layer-metadata/layer-metadata.composable.ts]: Indeed, I will still try to improve this with the external layers.

**Comment by tkohr** [comment]
Thanks for all the feedback @fgravin ! I've refactored the code as discussed, first into a composable and finally into a smarter displaying component that gets updated via a store and queries metadata via a service.

File: src/components/layer-metadata/layer-metadata.vue [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+import { layerMetadataService } from '@/services/layer-metadata/layer-metadata.service'
+import { LayerMetadataModel } from '@/services/layer-metadata/layer-metadata.model'
+
+const metadataStore = useMetadataStore()
+const { metadataTreeNode } = storeToRefs(metadataStore)
+const { t, i18next } = useTranslation()
+const layerMetadata: Ref<LayerMetadataModel | undefined> = ref()
+
+watch(metadataTreeNode, async node => {
+  node
```
Comment (fgravin) [inline src/components/layer-metadata/layer-metadata.vue:17]: weird syntax

`layerMetadata.value = node ? ... : ...` is more readable.

File: src/services/layer-metadata/layer-metadata.service.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+  async getLayerMetadata(
+    node: LayerTreeNodeModel,
+    currentLanguage: string
+  ): Promise<LayerMetadataModel> {
+    const layer: ThemeNodeModel | undefined = useThemes().findById(
+      parseInt(node.id, 10)
+    )
+    let localMetadata
+    let metadata
+    if (layer) {
```
Comment (fgravin) [inline src/services/layer-metadata/layer-metadata.service.ts:30]: Did you copy all this code from old code base ? It looks like old fashion

**Review by fgravin — APPROVED** [review]
Looks good.
It seems that I made some comments about the syntax which have been ignored. 
Also the metadata service is a bit heavy in term of coding style.
Nevermind

File: src/components/layer-metadata/layer-metadata.vue (hunk 10 above)
Comment (tkohr) [reply src/components/layer-metadata/layer-metadata.vue]: oups

File: src/services/layer-metadata/layer-metadata.service.ts (hunk 11 above)
Comment (tkohr) [reply src/services/layer-metadata/layer-metadata.service.ts:30]: Yes, I started from the old code, should've rather started from scratch in the end, I think. I'll have a look at this method again when integrating the WMTS.

**Pull request merged**